CC/src/DgmOctree.cpp

```diff
@@ -3494,7 +3494,15 @@ unsigned DgmOctree::executeFunctionForAllCellsAtLevel(	unsigned char level,
 		s_jumps = 0.0;
 		s_binarySearchCount = 0.0;
 #endif
-		CCParallelForEach(cells.begin(), cells.end(), LaunchOctreeCellFunc_MT);
+		if(maxThreadCount > 0)
+		{
+
+			CCParallelWithLimitedThreads(CCParallelForEach(cells.begin(), cells.end(), LaunchOctreeCellFunc_MT), maxThreadCount)
+		}
+		else
+		{
+			CCParallelForEach(cells.begin(), cells.end(), LaunchOctreeCellFunc_MT);
+		}
```

Comment:
I would be against this style _in general_. It's going to get messy quickly, makes it much easier to make mistakes, and is going to be a lot of "no fun" to maintain.

I don't think there's any measurable benefit to it.

If we need a mechanism to "breathe" in these functions, then we should be calling a user-specified generic "yield" function that gets called periodically. Then the user of the API can decide what to do - put up a progress dialog, or simply (in the Qt case) called QApplication::processEvents() to keep the UI responsive.

Or if you don't want to get that fancy, then simply call `std::this_thread::yield` periodically within the executing code which should avoid hogging all the processing.

Reply:
http://en.cppreference.com/w/cpp/thread/yield

Reply:
Then we still need an optional parameter to allow the system to breath or not?

Because some users will want to go as fast as possible, and they don't care that CC freezes the OS. But others will want to continue using their PC while the process is ongoing. In this case, calling QApplication::processEvents() only lets CC breath (but not the other applications). And it's hard to call "yield" on a regular basis (which is the only option to give a rather good experience to users wanting to continue working with their PC). If we call yield in the main octree fork for instance, the time between two calls would actually depend on the time spent on each cell, which is highly dependent on the number of cells (and the octree level by the way).

Am I missing something?

Reply:
I think we should maybe care to keep CC breath for example to allow a more friendly cancellation or a smoother monitoring of the progress but we should not take care of the OS, the OS scheduler should be smart enough for us.